File: CPP/Problem_124.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    const int N = 100000;
    const int K = 10000;

    vector<int> rad(N + 1, 1);

    for (int p = 2; p <= N; p++) {
        if (rad[p] == 1) { // prime
            for (int m = p; m <= N; m += p) {
                rad[m] *= p;
            }
        }
    }

    vector<pair<int,int>> v;
    v.reserve(N);
    for (int n = 1; n <= N; n++) v.push_back({rad[n], n});

    sort(v.begin(), v.end()); // (rad, n)
    cout << v[K - 1].second << "\n"; // 21417
    return 0;
}

File: Python/Problem_124.py
def solve(N: int = 100000, K: int = 10000) -> int:
    rad = [1] * (N + 1)

    for p in range(2, N + 1):
        if rad[p] == 1:  # prime
            for m in range(p, N + 1, p):
                rad[m] *= p

    pairs = [(rad[n], n) for n in range(1, N + 1)]
    pairs.sort()  # sorts by first then second automatically
    return pairs[K - 1][1]

if __name__ == "__main__":
    print(solve())  # 21417
